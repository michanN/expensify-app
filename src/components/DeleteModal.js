import React from 'react';
import Modal from 'react-modal';

const DeleteModal = props => (
  <Modal
    isOpen={!!props.showDeleteModal}
    onRequestClose={props.handleCloseDeleteModal}
    onRemove={props.onRemove}
    contentLabel="Delete Expense"
    closeTimeoutMS={200}
    className="modal"
  >
    <p className="modal__body">Are you sure you want to delete this expense?</p>
    <button
      className="button"
      onClick={props.onRemove}>
      Delete
    </button>
    <div className="modal-divider" />
    <button
      className="button button--delete"
      onClick={props.handleCloseDeleteModal}>
      Cancel
    </button>
  </Modal>
);

export default DeleteModal;
