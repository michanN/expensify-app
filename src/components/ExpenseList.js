import { connect } from 'react-redux';
import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
    {props.expenses.length === 0 ?
      (<p>No Expenses</p>) :
      (props.expenses.map((expense, index) => {
        return (<ExpenseListItem
          {...expense}
          key={index}
        />);
      }))
    }
  </div>
);

// connect returns a function that we can use to create the HOS
// function inside connect defines what we want our comp to access from the store
// store state gets passed in as the first argument
// When connecting a comp to the redux store the comp becomes reactive aka changes in the store,
// will rerender component

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
