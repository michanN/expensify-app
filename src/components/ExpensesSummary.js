import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expensesCount, expensesTotal, totalExpenseCount }) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
  const expensesTotalFormatted = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{expensesCount}</span> out of {totalExpenseCount} {expenseWord} totalling <span>{expensesTotalFormatted}</span> </h1>
        <div className="page-header__actions">
          <Link className="button button--summary" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  const numberOfExpenses = state.expenses.length;
  return {
    expensesCount: visibleExpenses.length,
    totalExpenseCount: numberOfExpenses,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
