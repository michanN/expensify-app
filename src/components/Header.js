import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogout } from '../actions/auth';

export const Header = props => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <div>
          <Link className="header__title" to="/dashboard">
            <h1>Expensify</h1>
          </Link>
        </div>
        <div>
          <span className="header__welcome">Hello {props.name}!</span>
          <button className="button button--link" onClick={props.startLogout}>Logout</button>
        </div>
      </div>
    </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = state => ({
  name: state.auth.name
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
