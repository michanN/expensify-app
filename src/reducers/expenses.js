const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD-EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates // Overrides every match
          };
        } else {
          return expense;
        };
      });
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id != action.id);
    default:
      return state;
  }
};

export default expensesReducer;
