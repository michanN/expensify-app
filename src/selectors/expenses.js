import moment from 'moment';

// GET VISIBLE EXPENSES

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const startDateMatch = startDate ? startDate.isSameOrBefore(expense.createdAt, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(expense.createdAt, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};

export default getVisibleExpenses;
