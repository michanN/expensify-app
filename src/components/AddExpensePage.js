import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        props.dispatch(addExpense(expense));
        props.history.push('/');
      }}
    />
  </div>
);

// If we only need dispatch and no state we can run connect without a function
export default connect()(AddExpensePage);
