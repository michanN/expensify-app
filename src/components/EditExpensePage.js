import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import DeleteModal from './DeleteModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showDeleteModal: false
    };
  }
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/dashboard');
  };
  onRemoveConfirm = () => {
    this.setState(() => ({
      showDeleteModal: false
    }));
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/dashboard');
  };
  onDeleteClick = () => {
    this.setState(() => ({
      showDeleteModal: true
    }));
  };
  onRemoveCancel = () => {
    this.setState(() => ({
      showDeleteModal: false
    }));
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button
            onClick={this.onDeleteClick}
            className="button button--delete"
            >
            Remove Expense
          </button>
        </div>
        <DeleteModal
          showDeleteModal={this.state.showDeleteModal}
          onRemoveCancel={this.onRemoveCancel}
          onRemoveConfirm={this.onRemoveConfirm}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
