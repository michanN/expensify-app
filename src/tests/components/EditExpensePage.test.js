import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import DeleteModal from '../../components/DeleteModal';
import expenses from '../fixtures/expenses';

let wrapper;
let startEditExpense;
let startRemoveExpense;
let history;

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage
    startEditExpense={startEditExpense}
    startRemoveExpense={startRemoveExpense}
    history={history}
    expense={expenses[0]}
  />);
});

test('should render EditExpensePage correclty', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/dashboard');
  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should handle onRemove', () => {
  wrapper.find('button').simulate('click');
  expect(wrapper.find(DeleteModal).prop('showDeleteModal')).toBeTruthy();
});
