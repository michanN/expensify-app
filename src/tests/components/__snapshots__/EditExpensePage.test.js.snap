// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correclty 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 150,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
